runtime/topk.cu

```diff
@@ -0,0 +1,166 @@
+// thread
+// - top_indices: Output array containing original indices of top-K values
+// - Each thread holds exactly ITEMS_PER_THREAD elements in both output arrays
+// - In the actual nvFuser-generated kernel, only the first k elements of each
+//   thread's output arrays should be used (via predication)
+// - This implementation sorts ALL elements and returns them in the output
+//   arrays, but the consuming kernel should only read the first k elements
+template <
+    int BLOCK_DIM_X,
+    int BLOCK_DIM_Y,
+    int BLOCK_DIM_Z,
+    int BLOCK_STATE_X,
+    int BLOCK_STATE_Y,
+    int BLOCK_STATE_Z,
+    typename DataT,
+    int ITEMS_PER_THREAD,
+    typename BlockDimT>
+__device__ void blockTopK(
+    DataT (&top_values)[ITEMS_PER_THREAD],
+    int64_t (&top_indices)[ITEMS_PER_THREAD],
+    const DataT (&input_data)[ITEMS_PER_THREAD],
+    int k,
+    bool largest,
+    bool sorted,
+    BlockDimT block_dim) {
+  // For now, only support all dimensions participating in sort (state=0)
+  // This follows the same constraint as argsort implementation
+  static_assert(
+      (isSort(BLOCK_STATE_X) || BLOCK_DIM_X == 1) &&
+          (isSort(BLOCK_STATE_Y) || BLOCK_DIM_Y == 1) &&
+          (isSort(BLOCK_STATE_Z) || BLOCK_DIM_Z == 1),
+      "For now, active TID dimensions must participate in sorting");
+
+  // Initialize output arrays directly (no temporary buffers needed)
+  for (int i = 0; i < ITEMS_PER_THREAD; i++) {
+    top_values[i] = CudaType<DataT>::get(input_data[i]);
+  }
+
+  // CUB BlockRadixSort setup - with proper multi-dimensional block support
+  // CUB supports multi-dimensional blocks when BLOCK_DIM_Y and BLOCK_DIM_Z are
+  // specified (identical to argsort configuration)
+  using BlockRadixSort = cub::BlockRadixSort<
+      typename CudaType<DataT>::type, // Key type
+      BLOCK_DIM_X, // X dimension
+      ITEMS_PER_THREAD, // Items per thread
+      int64_t, // Value type (for key-value sorting)
+      4, // RADIX_BITS (default)
+      true, // MEMOIZE_OUTER_SCAN (default for modern architectures)
+      cub::BLOCK_SCAN_WARP_SCANS, // INNER_SCAN_ALGORITHM (default)
+      cudaSharedMemBankSizeFourByte, // SMEM_CONFIG (default)
+      BLOCK_DIM_Y, // Y dimension
+      BLOCK_DIM_Z // Z dimension
+      >;
+
+  // Allocate shared memory for CUB operations
+  __shared__ typename BlockRadixSort::TempStorage temp_storage;
+
+  // Thread ID using all threads in the block (CUB requires full block
+  // participation) CUB doesn't support sorting of just blockDim.x, so all
+  // threads participate (identical to argsort pattern)
+  unsigned int thread_id =
+      index_utils::maskedOffset<true, true, true>(threadIdx, block_dim);
+
+  // Initialize indices array: 0, 1, 2, ..., n-1 per thread
+  // This creates the mapping between sorted values and their original positions
+  for (int i = 0; i < ITEMS_PER_THREAD; i++) {
+    top_indices[i] = thread_id * ITEMS_PER_THREAD + i;
```

Comment:
same comment at the place for initialization.

Reply:
Do you mean this has the same limitation as argsort? If that's the case, yes.